x/airdrop/keeper/claim_record.go

```diff
@@ -0,0 +1,240 @@
+		crs = append(crs, cr)
+		return false
+	})
+	return crs
+}
+
+// ClearClaimRecords deletes all the claim records of the given zone.
+func (k Keeper) ClearClaimRecords(ctx sdk.Context, chainId string) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte(chainId))
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
+		store.Delete(key)
+	}
+}
+
+// GetClaimableAmountForAction returns the amount claimable for the given
+// action, by the given address, against the given zone.
+func (k Keeper) GetClaimableAmountForAction(ctx sdk.Context, chainId string, address string, action types.Action) (uint64, error) {
+	cr, err := k.GetClaimRecord(ctx, chainId, address)
+	if err != nil {
+		return 0, err
+	}
+	if cr.Address == "" {
+		return 0, nil
+	}
+
+	// action completed, nothing to claim
+	if _, exists := cr.ActionCompleted[int32(action)]; exists {
+		return 0, nil
+	}
+
+	// get zone airdrop details
+	zd, ok := k.GetZoneDrop(ctx, chainId)
+	if !ok {
+		return 0, types.ErrZoneDropNotFound
+	}
+
+	// airdrop has not started
+	if ctx.BlockTime().Before(zd.StartTime) {
+		return 0, nil
+	}
+
+	// airdrop has already ended
+	if ctx.BlockTime().After(zd.StartTime.Add(zd.Duration).Add(zd.Decay)) {
+		return 0, nil
+	}
+
+	// calculate action allocation:
+	//   - zone drop action weight * claim record max allocation
+	amount := zd.Actions[int32(action)].MulInt64(int64(cr.MaxAllocation)).TruncateInt64()
+
+	// airdrop has not yet started to decay
+	if ctx.BlockTime().Before(zd.StartTime.Add(zd.Duration)) {
+		return uint64(amount), nil
+	}
+
+	// airdrop has started to decay, calculate claimable portion
+	elapsedDecayTime := ctx.BlockTime().Sub(zd.StartTime.Add(zd.Duration))
+	decayPercent := sdk.NewDec(elapsedDecayTime.Nanoseconds()).QuoInt64(zd.Decay.Nanoseconds())
+	claimablePercent := sdk.OneDec().Sub(decayPercent)
+	amount = claimablePercent.MulInt64(amount).TruncateInt64()
+
+	return uint64(amount), nil
+}
+
+// GetClaimableAmountForUser returns the amount claimable for the given user
+// against the given zone.
+func (k Keeper) GetClaimableAmountForUser(ctx sdk.Context, chainId string, address string) (uint64, error) {
+	cr, err := k.GetClaimRecord(ctx, chainId, address)
+	if err != nil {
+		return 0, err
+	}
+	if cr.Address == "" {
+		return 0, nil
+	}
+
+	// get zone airdrop details
+	zd, ok := k.GetZoneDrop(ctx, chainId)
+	if !ok {
+		return 0, types.ErrZoneDropNotFound
+	}
+
+	total := uint64(0)
+	// we will only need the index as we will be calling GetClaimableAmountForAction
+	for action := range zd.Actions {
+		claimableForAction, err := k.GetClaimableAmountForAction(ctx, chainId, address, types.Action(action))
+		if err != nil {
+			return 0, err
+		}
+		total += claimableForAction
+	}
+
+	return total, nil
+}
+
+// Claim
+func (k Keeper) Claim(ctx sdk.Context, chainId string, address string, action types.Action) (uint64, error) {
```

Comment:
This is only called post-validation, correct?

Reply:
We must ensure that it is called post-validation (on action)... yes